pennylane/transforms/resolve_dynamic_wires.py

```diff
@@ -0,0 +1,184 @@
+            return w, m.measurements
+
+        if self._dirty:
+            w = self._dirty.pop()
+            self._loaned[w] = "dirty"
+            return w, []
+        w = self._zeroed.pop()
+        self._loaned[w] = "zeroed"
+        return w, []
+
+    def return_wire(self, wire, reset_to_original=False):
+        """Return a wire label back to be re-used."""
+        reg_type = self._loaned.pop(wire)
+        if reg_type == "zeroed" and reset_to_original:
+            self._zeroed.append(wire)
+        else:
+            self._dirty.append(wire)
+
+
+def null_postprocessing(results: ResultBatch) -> Result:
+    """An empty postprocessing function returned by resolve_dynamic_wires"""
+    return results[0]
+
+
+@transform
+def resolve_dynamic_wires(
+    tape: QuantumScript,
+    zeroed: Sequence[Hashable] = (),
+    dirty: Sequence[Hashable] = (),
+    min_integer: Optional[int] = None,
+) -> tuple[QuantumScriptBatch, PostprocessingFn]:
+    """Map dynamic wires to concrete values determined by the provided ``zeroed`` and ``dirty`` registers.
+
+    Args:
+        tape (QuantumScript): A circuit that may contain dynamic wire allocations and deallocations
+        zeroed (Sequence[Hashable]): a register of wires known to be the zero state
+        dirty (Sequence[Hashable]): a register of wires with any state
+
+    Returns:
+        tuple[QuantumScript], Callable[[ResultBatch], Result]: A batch of tapes and a postprocessing function
+
+    .. code-block:: python
+
+        @qml.qnode(qml.device('default.qubit'))
+        def circuit(, require_zeros=True):
+            with qml.allocation.safe_allocate(1, require_zeros=require_zeros) as wires:
+                qml.X(wires)
+            with qml.allocation.safe_allocate(1, require_zeros=require_zeros) as wires:
+                qml.Y(wires)
+            return qml.state()
+
+    >>> print(qml.draw(circuit)())
+    <DynamicWire>: ──Allocate──X──Deallocate─┤  State
+    <DynamicWire>: ──Allocate──Y──Deallocate─┤  State
+
+    If we provide two zeroed qubits to the transform, we can see that the two operations have been
+    assigned to both wires known to be in the zero state.
+
+    >>> assigned_two_zeroed = qml.transforms.resolve_dynamic_wires(circuit, zeroed=("a", "b"))
+    >>> print(qml.draw(assigned_two_zeroed)())
+    a: ──Y─┤  State
+    b: ──X─┤  State
+
+    If we only provide one zeroed wire, we perform a reset on that wire before reusing for the ``Y`` operation.
+
+    >>> assigned_one_zeroed = qml.transforms.resolve_dynamic_wires(circuit, zeroed=("a", ))
+    >>> print(qml.draw(assigned_one_zeroed)())
+    a: ──X──┤↗│  │0⟩──Y─┤  State
+
+    If we only provide dirty qubits with unknown states, then they will be reset to zero before being used
+    in an operation that requires a zero state.
+
+    >>> assigned_dirty = qml.transforms.resolve_dynamic_wires(circuit, dirty=("a", "b"))
+    >>> print(qml.draw(assigned_dirty)())
+    b: ──┤↗│  │0⟩──X──┤↗│  │0⟩──Y─┤  State
```

Comment:
I suppose there is a choice to be made here: Do you use one wire and reset it twice (initially and between the ops), or do you reset both wires initially and use the both (essentially a depth vs width tradeoff).
If a user passes those two wires, maybe it's nice to use them both instead of increasing depth :thinking: 

The fact that wire `"b"` is chosen is a bit unexpected from a user perspective I think (`pop` is the culprit I believe?)